[Reuben Thomas] iP

reuben-thomas opened the pull request:
# 📝 KIPP, your spaceflight task assistant

> "Do not go gentle into that good night. Old age should burn and rave at close of ay. Rage, rage against the dying of
> the light." - Dylan Thomas

## Overview

Inspired by the robot in Interstellar, KIPP is meant to be a skilled, useful machine in your life, allowing you to keep track of your tasks as you try to save humanity. At the time being, it allows you to add the following three categories of tasks

- ✅ **Todo**: Tasks that are not timed
- ⏳ **Deadline**: Tasks that must be completed by a deadline
- 📅 **Event**: Tasks or events within a start and end time range

## Usage

In order to execute the CLI based KIPP chat interface, simply create a chat object and run as follows:

``` Java
import kippchat.KippChat;

public static void main(String[] args) {
    KippChat chat = new KippChat();
    chat.run();
}
```

## Project Structure

The project is abstracted into the following packages, and their functions are as follows:

- `/commandhandler`: Create standardized commands and handler functions, as well as automatic fallback behaviour and automatically generated help commands
- `/kipp`: Define functions and KIPP's logic interacting with storage and tasklist
- `/kippchat`: CLI-based UI and entry point for the program
- `/storage`: Store and retrieve persistent data to disk
- `/tasklist`: Tasklist management, and types of tasks

``` pgsql
📦 KIPP
 ├── ...
 │   ├── 📂 main/java  
 │   │   ├── 📂 commandhandler 
 │   │   ├── 📂 kipp
 │   │   ├── 📂 kippchat 
 │   │   ├── 📂 storage
 │   │   ├── 📂 tasklist 
 │   ├── ...
 ├── ...
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

reuben-thomas committed:
Update name to Kipp, add basic greet and exit functionality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Adding simple echo functionality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Adding functionality to add and list to-do items

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add function to mark tasks complete or incomplete.

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add todo, deadline, event task types.

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add references section in README.md

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Fix basic test cases

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add input validation and error handling

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add delete functionality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add load and save functionality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-Level-7'

reuben-thomas committed:
Rename ToDoList to TaskManager

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Switch to using LocalDate for tasks

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Load save file by default

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-Level-8'

reuben-thomas committed:
Rename TaskManager to TaskList

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add storage class and new exception

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add command handler as separate class

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'A-MoreOOP'

reuben-thomas committed:
Refactor by packages

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'A-Packages'

reuben-thomas committed:
Merge remote-tracking branch 'origin/add-gradle-support'

reuben-thomas committed:
Configure gradle to execute Kipp

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'A-Gradle'

reuben-thomas committed:
Add JUnit tests for Storage and TaskList

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-A-JUnit'

reuben-thomas committed:
Add test for TaskList equality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Rename output for Jar plugin

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add javadoc comments

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Conform to coding standard

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add functionality to find tasks

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add checkstyle

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Move chat to package

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-A-JavaDoc'

reuben-thomas committed:
Merge branch 'branch-A-CodingStandard'

reuben-thomas committed:
Merge branch 'branch-Level-9'

dexterkwxn commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,83 @@
+     */
+    public void save(T item) throws StorageException {
+        this.ensureFileExists();
+        try (FileOutputStream fileOut = new FileOutputStream(this.relativeFilePath); ObjectOutputStream objectOut = new ObjectOutputStream(fileOut)) {
```
This line seems a little long, maybe consider splitting it up?
```suggestion
        try (
            FileOutputStream fileOut = new FileOutputStream(this.relativeFilePath);
            ObjectOutputStream objectOut = new ObjectOutputStream(fileOut)
        ) {
```

dexterkwxn commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,83 @@
+     */
+    public T load() throws StorageException {
+        this.ensureFileExists();
+        try (FileInputStream fileIn = new FileInputStream(this.relativeFilePath); ObjectInputStream objectIn = new ObjectInputStream(fileIn)) {
```
Similar to above.

dexterkwxn commented on src/test/java/storage/StorageTest.java:
```diff
@@ -0,0 +1,77 @@
+            taskListStorage.load();
+            fail();
+        } catch (StorageException e) {
+            assertEquals(String.format("Error loading data from file: %s.", StorageTest.TEST_FILE_RELATIVE_PATH), e.getMessage(), "Loading a file with an invalid type should throw an exception.");
```
Line seems really long here as well.
```suggestion
            assertEquals(
                String.format("Error loading data from file: %s.", StorageTest.TEST_FILE_RELATIVE_PATH), 
                e.getMessage(), 
                "Loading a file with an invalid type should throw an exception."
            );
```

dexterkwxn commented on src/main/java/kipp/Kipp.java:
```diff
@@ -0,0 +1,315 @@
+
+public class Kipp {
+    private static final String LOGO = """
+            ██   ██ ██ ██████  ██████
+            ██  ██  ██ ██   ██ ██   ██
+            █████   ██ ██████  ██████
+            ██  ██  ██ ██      ██
+            ██   ██ ██ ██      ██
```
Nice interstellar reference ;)

dexterkwxn reviewed (commented):
LGTM. Some small nits with long lines, but otherwise looks fine :)

reuben-thomas committed:
Add image and quote to README

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Remove prior checkstyle

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add and address checkstyle

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Create gradle.yml

reuben-thomas committed:
Merge branch 'branch-A-CheckStyle'

reuben-thomas committed:
Add JavaFx gui

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add exit functionality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-Level-10'

reuben-thomas committed:
Enable JUnit test

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Fix pause after goodbye message before program exit

When provided a bye command, the application is expected to display the
bot's goodbye message for three seconds prior to closing the window. The
application pauses and exits without showing the goodbye message.

This occurs due to the use of Thread.sleep to create a pause in
execution, which puts the thread to sleep before the UI updates, and the
call to System.exit immediately after. Thus, a different wait mechanism
is necessary.

Let's use the PauseTransition within JavaFX, which allows us to pause
the program without blocking the UI thread.

Let's also use Platform.exit to close the application instead, as it
will guarantee that in the future, any UI updates can be completed
before closing.

reuben-thomas committed:
Use gradle check, and test across multiple platforms in workflow

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Only create pause on exit for goodbye command

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add assertions to storage package to avoid invalid states

The storage package allows users to instantiate an instance of
storage.Storage with a null type, empty filepath, and even save a null
object to a file.

This can lead to invalid states, since type information is required for
typecasting when loading from a file. Furthermore, saving a null object
should not be allowed as the user should be forced to delete the file
instead. Lastly, an empty filepath should not be allowed as it does not
provide a valid file to save to.

Let's make use of assertions to enforce the aforementioned constraints.

We choose to use assertions rather than exceptions for these cases, as
these are invalid states that the implementer should not allow to
happen, rather than exceptional errors to be handled.

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Replace public constructors with factory methods

The codebase makes use of public extensively, which is often regarded as
bad practice and less maintainable, since the code associated with the
construction of an object, and the code used to interface with the
object are the same.

Thus, if the constructor needs to be changed in the future, as is
likely, all instances of its usage in other files must also be changed,
which makes it less maintainable.

Let's implement factory methods associated with the creation of objects,
that can be called from other classes. Furthermore, let's make the
constructors private so that they may not accidentally be used to
instantiate objects.

Factory methods are chosen instead of the builder pattern, since it
allows us to implement it quicker, and for less additional code.

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add JavaDoc comments for factory methods

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Assert non-empty response message in CommandResult

The CommandResult class does not enforce on instantiation, whether the
response or error optional strings should be empty, and if present, that
they are not blank.

Let's add assertions to the constructor to enforce that for results
associated with invalid usage, and unexpected errors must have a
non-empty error message and an empty response message. Likewise, for a
result associated with successful execution, the response must include a
non-empty response message and an empty error message.

We choose to enforce this through assertions rather than throwing an
exception, since this is a case of incorrect functioning or modification
of the class.

In the future, a factory method may be used to further guarantee as the
only exposed methods to instantiate a CommandResult, while the
constructor continues to manage the current assertions implemented.

reuben-thomas committed:
Merge pull request #2 from reuben-thomas/branch-A-CodeQuality

Replace constructors with factory methods

reuben-thomas committed:
Merge branch 'master' into branch-A-Assertions

reuben-thomas committed:
Add assertions to enforce unique command names

A user should not be able to add multiple commands with the same name.
The CommandHandler class method to add a command simply replaces the
existing command with the new one without any checks or indication to
the user.

Let's make use of assertions to enforce that the new command being added
is unique, and does not already exist in the command map.

We make use of assertions instead of exceptions in this case, since it
is a case of incorrect usage of the class by a programmer when defining
a set of commands associated with a command handler prior to execution.
It is critical to prevent existing commands from being mistakenly
overwritten.

reuben-thomas committed:
Merge pull request #1 from reuben-thomas/branch-A-Assertions

Add assertions to handle invalid states

reuben-thomas committed:
Add load and save commands from specified file

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-C-FlexibleDataSource'

reuben-thomas committed:
Update gitignore

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Improve code quality abiding by SLAP and shortening functions

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Improve JavaDoc comments

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Remove outdated save file path

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add personality

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-A-Personality'

reuben-thomas committed:
Allow resizing elements in GUI

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Set theme as space terminal

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add chat and dialog background

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Standardize UI fonts

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-A-BetterGui'

reuben-thomas committed:
Add updated UI screenshot

reuben-thomas committed:
Add command summary to docs

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Add quick start to user guide

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Merge branch 'branch-A-UserGuide'

reuben-thomas committed:
Update image references

Signed-off-by: Reuben Thomas <reubenthomas123@gmail.com>

reuben-thomas committed:
Update screenshot image preview method

reuben-thomas committed:
Remove div from image preview

reuben-thomas committed:
Standardize font size across Ui

reuben-thomas committed:
Update Ui screenshot
